File: storeorder.py
#store order management 

from typing import List, Dict, Optional, Tuple, Union
from dataclasses import dataclass
from datetime import datetime
import time
from collections import deque

@dataclass
class Customer:
    id: int
    name: str
    loyalty_points: int = 0

@dataclass
class Product:
    id: int
    name: str
    price: float
    stock: int

@dataclass
class OrderItem:
    product_id: int
    quantity: int

@dataclass
class Order:
    id: int
    customer_id: int
    products: List[OrderItem]
    status: str
    timestamp: datetime

@dataclass
class HistoryEntry:
    order_id: int
    status: str
    timestamp: datetime

class Store:
    def __init__(self):
        self.customers: List[Customer] = []
        self.products: Dict[int, Product] = {}
        self.orders: List[Order] = []
        self.order_queue: deque = deque()
        self.priority_orders: List[int] = []
        self.order_history: List[HistoryEntry] = []
        
    def add_customer(self, name: str) -> int:
        customer_id = len(self.customers) + 1
        self.customers.append(Customer(id=customer_id, name=name))
        return customer_id
    
    def find_customer(self, customer_id: int) -> Optional[Customer]:
        for customer in self.customers:
            if customer.id == customer_id:
                return customer
        return None
    
    def add_product(self, name: str, price: float, initial_stock: int) -> int:
        product_id = len(self.products) + 1
        self.products[product_id] = Product(
            id=product_id, 
            name=name, 
            price=price, 
            stock=initial_stock
        )
        return product_id
    
    def update_stock(self, product_id: int, quantity: int) -> bool:
        if product_id not in self.products:
            return False
        
        self.products[product_id].stock += quantity
        return True
    
    def create_order(self, customer_id: int, products: List[OrderItem]) -> Optional[int]:
        customer = self.find_customer(customer_id)
        if not customer:
            return None
            
        for item in products:
            if (item.product_id not in self.products or 
                self.products[item.product_id].stock < item.quantity):
                return None
        
        order_id = len(self.orders) + 1
        order = Order(
            id=order_id,
            customer_id=customer_id,
            products=products,
            status='pending',
            timestamp=datetime.now()
        )
        
        self.orders.append(order)
        
        if customer.loyalty_points >= 100:
            self.priority_orders.append(order_id)
        else:
            self.order_queue.append(order_id)
        
        self.order_history.append(HistoryEntry(
            order_id=order_id,
            status='pending',
            timestamp=datetime.now()
        ))
        
        for item in products:
            self.products[item.product_id].stock -= item.quantity
        
        total_spent = sum(
            self.products[item.product_id].price * item.quantity
            for item in products
        )
        
        customer.loyalty_points += int(total_spent)
        
        return order_id
    
    def process_next_order(self) -> Optional[Order]:
        if self.priority_orders:
            order_id = self.priority_orders.pop()
            return self._process_order(order_id)
        
        if self.order_queue:
            order_id = self.order_queue.popleft()
            return self._process_order(order_id)
        
        return None
    
    def _process_order(self, order_id: int) -> Optional[Order]:
        order = next((order for order in self.orders if order.id == order_id), None)
        if not order:
            return None
        
        order.status = 'processing'
        
        self.order_history.append(HistoryEntry(
            order_id=order.id,
            status='processing',
            timestamp=datetime.now()
        ))
        
        def complete_order():
            time.sleep(2)
            order.status = 'completed'
            
            self.order_history.append(HistoryEntry(
                order_id=order.id,
                status='completed',
                timestamp=datetime.now()
            ))
        
        order.status = 'completed'
        self.order_history.append(HistoryEntry(
            order_id=order.id,
            status='completed',
            timestamp=datetime.now()
        ))
        
        return order
    
    def cancel_order(self, order_id: int) -> bool:
        order = next((order for order in self.orders if order.id == order_id), None)
        if not order:
            return False
        
        if order.status not in ['pending', 'processing']:
            return False
        
        order.status = 'cancelled'
        
        self.order_history.append(HistoryEntry(
            order_id=order_id,
            status='cancelled',
            timestamp=datetime.now()
        ))
        
        for item in order.products:
            self.update_stock(item.product_id, item.quantity)
        
        if order_id in self.order_queue:
            queue_list = list(self.order_queue)
            queue_list.remove(order_id)
            self.order_queue = deque(queue_list)
            
        if order_id in self.priority_orders:
            self.priority_orders.remove(order_id)
        
        return True
    
    def get_order_history(self, order_id: int) -> List[Dict[str, Union[str, datetime]]]:
        return [
            {"status": entry.status, "timestamp": entry.timestamp}
            for entry in self.order_history
            if entry.order_id == order_id
        ]
    
    def get_backlog_summary(self) -> Dict[str, int]:
        return {
            "regular": len(self.order_queue),
            "priority": len(self.priority_orders)
        }

def run_store_simulation():
    store = Store()
    
    customer1 = store.add_customer("Alice")
    customer2 = store.add_customer("Bob")
    
    laptop = store.add_product("Laptop", 999.99, 5)
    phone = store.add_product("Smartphone", 499.99, 10)
    headphones = store.add_product("Wireless Headphones", 129.99, 15)
    
    order1 = store.create_order(customer1, [
        OrderItem(product_id=laptop, quantity=1),
        OrderItem(product_id=headphones, quantity=1)
    ])
    
    customer = store.find_customer(customer2)
    if customer:
        customer.loyalty_points = 150
    
    order2 = store.create_order(customer2, [
        OrderItem(product_id=phone, quantity=2)
    ])
    
    print("Processing orders...")
    processed1 = store.process_next_order()
    if processed1:
        print(f"Processed order: {processed1.id}")
    
    backlog = store.get_backlog_summary()
    print(f"Orders in queue: Regular={backlog['regular']}, Priority={backlog['priority']}")
    
    processed2 = store.process_next_order()
    if processed2:
        print(f"Processed order: {processed2.id}")
    
    if order1:
        history = store.get_order_history(order1)
        print(f"Order {order1} history:", history)

if __name__ == "__main__":
    run_store_simulation()

File: storeorder.ts
//Store Order Management System

interface Customer {
  id: number;
  name: string;
  loyaltyPoints: number;
}

interface Product {
  id: number;
  name: string;
  price: number;
  stock: number;
}

interface Order {
  id: number;
  customerId: number;
  products: Array<{ productId: number; quantity: number }>;
  status: 'pending' | 'processing' | 'completed' | 'cancelled';
  timestamp: Date;
}

class Store {
  private customers: Customer[] = [];
  private products: Map<number, Product> = new Map();
  private orders: Order[] = [];
  private orderQueue: number[] = [];
  private priorityOrders: number[] = [];
  private orderHistory: { orderId: number, status: string, timestamp: Date }[] = [];

  constructor() {}

  addCustomer(name: string): number {
    const id = this.customers.length + 1;
    this.customers.push({ id, name, loyaltyPoints: 0 });
    return id;
  }

  findCustomer(id: number): Customer | undefined {
    return this.customers.find(customer => customer.id === id);
  }

  addProduct(name: string, price: number, initialStock: number): number {
    const id = this.products.size + 1;
    this.products.set(id, { id, name, price, stock: initialStock });
    return id;
  }

  updateStock(productId: number, quantity: number): boolean {
    const product = this.products.get(productId);
    if (!product) return false;
    
    product.stock += quantity;
    this.products.set(productId, product);
    return true;
  }

  createOrder(customerId: number, products: Array<{ productId: number; quantity: number }>): number | null {
    const customer = this.findCustomer(customerId);
    if (!customer) return null;

    for (const item of products) {
      const product = this.products.get(item.productId);
      if (!product || product.stock < item.quantity) {
        return null;
      }
    }

    const orderId = this.orders.length + 1;
    const order: Order = {
      id: orderId,
      customerId,
      products,
      status: 'pending',
      timestamp: new Date()
    };
    
    this.orders.push(order);
    
    if (customer.loyaltyPoints >= 100) {
      this.priorityOrders.push(orderId);
    } else {
      this.orderQueue.push(orderId);
    }
    
    this.orderHistory.push({
      orderId,
      status: 'pending',
      timestamp: new Date()
    });
    
    for (const item of products) {
      const product = this.products.get(item.productId)!;
      product.stock -= item.quantity;
      this.products.set(item.productId, product);
    }
    
    const totalSpent = products.reduce((sum, item) => {
      const product = this.products.get(item.productId)!;
      return sum + (product.price * item.quantity);
    }, 0);
    
    customer.loyaltyPoints += Math.floor(totalSpent);
    
    return orderId;
  }

  processNextOrder(): Order | null {
    if (this.priorityOrders.length > 0) {
      const orderId = this.priorityOrders.pop()!;
      return this.processOrder(orderId);
    }
    
    if (this.orderQueue.length > 0) {
      const orderId = this.orderQueue.shift()!;
      return this.processOrder(orderId);
    }
    
    return null;
  }

  private processOrder(orderId: number): Order | null {
    const orderIndex = this.orders.findIndex(order => order.id === orderId);
    if (orderIndex === -1) return null;
    
    const order = this.orders[orderIndex];
    order.status = 'processing';
    
    this.orderHistory.push({
      orderId: order.id,
      status: 'processing',
      timestamp: new Date()
    });
    
    setTimeout(() => {
      order.status = 'completed';
      
      this.orderHistory.push({
        orderId: order.id,
        status: 'completed',
        timestamp: new Date()
      });
    }, 2000);
    
    return order;
  }

  cancelOrder(orderId: number): boolean {
    const orderIndex = this.orders.findIndex(order => order.id === orderId);
    if (orderIndex === -1) return false;
    
    const order = this.orders[orderIndex];
    if (order.status !== 'pending' && order.status !== 'processing') {
      return false;
    }
    
    order.status = 'cancelled';
    
    this.orderHistory.push({
      orderId,
      status: 'cancelled',
      timestamp: new Date()
    });
    
    for (const item of order.products) {
      this.updateStock(item.productId, item.quantity);
    }
    
    const queueIndex = this.orderQueue.indexOf(orderId);
    if (queueIndex !== -1) {
      this.orderQueue.splice(queueIndex, 1);
    }
    
    const priorityIndex = this.priorityOrders.indexOf(orderId);
    if (priorityIndex !== -1) {
      this.priorityOrders.splice(priorityIndex, 1);
    }
    
    return true;
  }

  getOrderHistory(orderId: number): { status: string, timestamp: Date }[] {
    return this.orderHistory
      .filter(entry => entry.orderId === orderId)
      .map(({ status, timestamp }) => ({ status, timestamp }));
  }

  getBacklogSummary(): { regular: number, priority: number } {
    return {
      regular: this.orderQueue.length,
      priority: this.priorityOrders.length
    };
  }
}

function runStoreSimulation() {
  const store = new Store();
  
  const customer1 = store.addCustomer("Alice");
  const customer2 = store.addCustomer("Bob");
  
  const laptop = store.addProduct("Laptop", 999.99, 5);
  const phone = store.addProduct("Smartphone", 499.99, 10);
  const headphones = store.addProduct("Wireless Headphones", 129.99, 15);
  
  const order1 = store.createOrder(customer1, [
    { productId: laptop, quantity: 1 },
    { productId: headphones, quantity: 1 }
  ]);
  
  store.findCustomer(customer2)!.loyaltyPoints = 150;
  const order2 = store.createOrder(customer2, [
    { productId: phone, quantity: 2 }
  ]);
  
  console.log("Processing orders...");
  const processed1 = store.processNextOrder();
  console.log(`Processed order: ${processed1?.id}`);
  
  const backlog = store.getBacklogSummary();
  console.log(`Orders in queue: Regular=${backlog.regular}, Priority=${backlog.priority}`);
  
  const processed2 = store.processNextOrder();
  console.log(`Processed order: ${processed2?.id}`);
  
  if (order1) {
    const history = store.getOrderHistory(order1);
    console.log(`Order ${order1} history:`, history);
  }
}

runStoreSimulation();
